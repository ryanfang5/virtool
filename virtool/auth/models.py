from sqlalchemy import Column, Enum, String

from virtool.pg.base import Base
from virtool.pg.utils import SQLEnum


class ResourceType(str, SQLEnum):
    """
    Types of resources a permission can apply to (e.g. 'app', 'sample', 'group').

    """

    app = "app"
    group = "group"


class Action(str, SQLEnum):
    """
    Conserved actions that can be performed on a resource and
    are controlled by permissions.
    """

    create = "create"
    update = "update"
    delete = "delete"
    modify = "modify"
    remove = "remove"


class SQLPermission(Base):
    """
    SQL model to store permissions

    """

    __tablename__ = "permissions"

    id = Column(String, primary_key=True)
    name = Column(String)
    description = Column(String)
    resource_type = Column(Enum(ResourceType))
    action = Column(Enum(Action))
