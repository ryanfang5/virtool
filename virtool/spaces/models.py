from sqlalchemy import (
    Column,
    DateTime,
    Integer,
    String,
)

from virtool.pg.base import Base


class SpaceModel(Base):
    __tablename__ = "space"

    id = Column(Integer, primary_key=True)
    name = Column(String, unique=True)
    description = Column(String)
    created_at = Column(DateTime, nullable=False)
    updated_at = Column(DateTime, nullable=False)
    created_by = Column(String, nullable=False)

