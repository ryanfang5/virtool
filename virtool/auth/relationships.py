class BaseRelationship:

    def __init__(self, relation):
        self.object_name = "virtool"
        self.object_type = "app"
        self.relation = relation

    ...


class GroupPermission(BaseRelationship):

    def __init__(self, group_id, relation):
        super().__init__(relation)
        self.user_id = group_id
        self.user_type = "group"


class UserPermission(BaseRelationship):

    def __init__(self, user_id, relation):
        super().__init__(relation)
        self.user_id = user_id
        self.user_type = "user"

