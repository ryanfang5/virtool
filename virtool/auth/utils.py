import sys
from enum import Enum
from logging import getLogger

import openfga_sdk
from aiohttp import ClientConnectorError
from openfga_sdk import (
    CreateStoreRequest,
    WriteAuthorizationModelRequest,
    TypeDefinition,
    Userset,
    OpenFgaApi,
    WriteRequest,
    TupleKey,
    ApiException,
    TupleKeys,
)
from openfga_sdk.api import open_fga_api

from virtool.auth.permissions import (
    AppPermission,
)

logger = getLogger("openfga")


async def connect_openfga(openfga_host: str, openfga_scheme: str):
    """
    Connects to an OpenFGA server and configures the store id.
    Returns the application client instance.
    """
    configuration = openfga_sdk.Configuration(
        api_scheme=openfga_scheme, api_host=openfga_host
    )

    logger.info("Connecting to OpenFGA")

    try:
        api_client = openfga_sdk.ApiClient(configuration)

        api_instance = open_fga_api.OpenFgaApi(api_client)

        configuration.store_id = await get_or_create_store(api_instance)

        await write_auth_model(api_instance)

    except ClientConnectorError:
        logger.critical("Could not connect")
        sys.exit(1)

    return api_instance


async def get_or_create_store(api_instance: OpenFgaApi):
    """
    Get the OpenFGA Store or create one if it does not exist.

    :return: the store id
    """
    response = await api_instance.list_stores()

    logger.info("Connected to OpenFGA")

    if response.stores:
        logger.info("Found existing store")
        return response.stores[0].id

    body = CreateStoreRequest(
        name="Virtool",
    )
    logger.info("Creating store")
    response = await api_instance.create_store(body)
    return response.id


async def write_auth_model(api_instance: OpenFgaApi):
    """
    Write the authorization model for the OpenFGA Store if it does not exist.
    """
    response = await api_instance.read_authorization_models()

    app_definition = TypeDefinition(
        type="app",
        relations={permission.name: Userset(this={}) for permission in AppPermission},
    )

    group_definition = TypeDefinition(
        type="group",
        relations={"member": Userset(this={})},
    )

    type_definitions = WriteAuthorizationModelRequest(
        type_definitions=[app_definition, group_definition],
    )

    if (
        response.authorization_models
        and response.authorization_models[0].type_definitions
        == type_definitions.type_definitions
    ):
        return

    await api_instance.write_authorization_model(type_definitions)


async def write_tuple(
    api_instance, user_type, user_id, relations, object_type, object_name
):
    """
    Write a relationship tuple in OpenFGA.
    """

    tuple_list = [
        TupleKey(
            user=f"{user_type}:{user_id}",
            relation=relation.name if isinstance(relation, Enum) else relation,
            object=f"{object_type}:{object_name}",
        )
        for relation in relations
    ]

    body = WriteRequest(writes=TupleKeys(tuple_keys=tuple_list))

    try:
        await api_instance.write(body)
    except ApiException:
        pass
