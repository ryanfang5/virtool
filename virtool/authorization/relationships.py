from abc import ABC, abstractmethod
from typing import Union

from virtool.authorization.permissions import PermissionType, ResourceType


class AbstractRelationship(ABC):
    @property
    @abstractmethod
    def object_id(self) -> Union[str, int]:
        ...

    @property
    @abstractmethod
    def object_type(self) -> ResourceType:
        ...

    @property
    @abstractmethod
    def user_id(self) -> str:
        ...

    @property
    @abstractmethod
    def user_type(self) -> str:
        ...

    @property
    @abstractmethod
    def relation(self) -> str:
        ...


class GroupMembership(AbstractRelationship):
    def __init__(self, user_id, group_id):
        self._user_id = user_id
        self._group_id = group_id

    @property
    def relation(self):
        return "member"

    @property
    def object_id(self) -> str:
        return self._group_id

    @property
    def object_type(self) -> str:
        return ResourceType.GROUP.value

    @property
    def user_id(self) -> str:
        return self._user_id

    @property
    def user_type(self) -> str:
        return "user"


class GroupPermission(AbstractRelationship):
    def __init__(
        self,
        group_id,
        permission: PermissionType,
    ):
        self._group_id = group_id
        self._permission = permission

    @property
    def object_id(self) -> str:
        return 0

    @property
    def object_type(self) -> str:
        return ResourceType.SPACE.value

    @property
    def relation(self) -> str:
        return self._permission.value.id

    @property
    def user_id(self) -> str:
        return f"{self._group_id}#member"

    @property
    def user_type(self) -> str:
        return "group"


class UserPermission(AbstractRelationship):
    def __init__(self, user_id: Union[int, str], permission: PermissionType):
        self._user_id = user_id
        self._permission = permission

    def __repr__(self):
        return f"UserPermission(user_id={self._user_id}, permission={self._permission}, object_id={self.object_id}, object_type={self.object_type})"

    @property
    def object_id(self) -> Union[int, str]:
        return 0

    @property
    def object_type(self) -> str:
        return ResourceType.SPACE.value

    @property
    def relation(self) -> str:
        return self._permission.value.id

    @property
    def user_id(self) -> Union[int, str]:
        return self._user_id

    @property
    def user_type(self) -> str:
        return "user"
