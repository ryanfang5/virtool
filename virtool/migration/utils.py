from pathlib import Path


def get_revisions_path() -> Path:
    """
    Get the hardcoded path to data revision files.

    :return: the revisions path

    """
    return Path(__file__).parent.parent.parent / "assets/revisions"
