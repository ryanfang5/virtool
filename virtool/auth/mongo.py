from typing import List

from virtool.auth.permissions import PermissionType
from virtool.data.errors import ResourceNotFoundError
from virtool.mongo.core import DB
from virtool.mongo.utils import get_one_field


async def check_in_mongo(mongo: DB, user_id: str, permission: PermissionType) -> bool:
    """
    Check permissions in Mongo.
    """

    return (
        await mongo.users.count_documents(
            {"_id": user_id, f"permissions.{permission.name}": True}, limit=1
        )
        == 1
    )


async def list_permissions_in_mongo(mongo: DB, user_id: str) -> List[PermissionType]:
    """
    List user permissions in Mongo.
    """

    permissions = await get_one_field(mongo.users, "permissions", user_id)

    if permissions:
        return sorted(
            [permission for permission in permissions if permissions[permission]]
        )

    raise ResourceNotFoundError()
