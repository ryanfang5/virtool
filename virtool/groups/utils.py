from __future__ import annotations

from copy import deepcopy
from typing import List, Dict, Tuple

import virtool.users.utils
from virtool.authorization.permissions import SpacePermission
from virtool.authorization.relationships import GroupPermission
from virtool.users.utils import Permission


def merge_group_permissions(groups: List[dict]) -> Dict[str, bool]:
    """
    Return a :class:`dict` of permissions that will be
    inherited by a user belonging to all the passed ``groups``.

    :param groups: a list of group documents.
    :return: a dict keyed by permission names with boolean values
    indicating the state of the permission

    """
    permission_dict = virtool.users.utils.generate_base_permissions()

    for p in Permission:
        for group in groups:
            try:
                if group["permissions"][p.value]:
                    permission_dict[p.value] = True
            except KeyError:
                permission_dict[p.value] = False

    return permission_dict


def adapt_legacy_permissions_dict(permissions_dict: dict[str, bool]):
    adapted = deepcopy(permissions_dict)

    adapted.pop("remove_job", None)

    adapted["create_reference"] = adapted.pop("create_ref")
    adapted["delete_upload"] = adapted.pop("remove_file")
    adapted["create_upload"] = adapted.pop("upload_file")

    return adapted


def convert_permissions_dict_to_relationships(
    group_id: str,
    permissions_dict: dict[str, bool],
) -> Tuple[List[GroupPermission], List[GroupPermission]]:
    adapted_permissions_dict = adapt_legacy_permissions_dict(permissions_dict)

    adds = [
        GroupPermission(group_id, SpacePermission.from_string(permission))
        for permission, value in adapted_permissions_dict.items()
        if value
    ]

    removes = [
        GroupPermission(group_id, SpacePermission.from_string(permission))
        for permission, value in adapted_permissions_dict.items()
        if not value
    ]

    return adds, removes
