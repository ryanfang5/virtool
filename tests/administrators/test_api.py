import pytest
from virtool_core.models.roles import AdministratorRole

from virtool.authorization.relationships import AdministratorRoleAssignment
from virtool.mongo.utils import get_one_field


@pytest.mark.apitest
async def test_list_administrators(spawn_client, fake2, snapshot):
    client = await spawn_client(authorize=True, administrator=True)

    user_1 = await fake2.users.create()

    user_2 = await fake2.users.create()

    authorization_client = client.app["authorization"]

    await authorization_client.add(
        AdministratorRoleAssignment(user_1.id, AdministratorRole.BASE),
        AdministratorRoleAssignment(user_2.id, AdministratorRole.FULL),
    )

    resp = await client.get("/administrators")

    assert resp.status == 200

    assert await resp.json() == snapshot


@pytest.mark.apitest
async def test_get_administrator(spawn_client, fake2, snapshot):
    client = await spawn_client(authorize=True, administrator=True)

    user = await fake2.users.create()

    authorization_client = client.app["authorization"]

    await authorization_client.add(
        AdministratorRoleAssignment(user.id, AdministratorRole.BASE),
    )

    resp = await client.get(f"/administrators/{user.id}")

    assert resp.status == 200

    assert await resp.json() == snapshot


@pytest.mark.apitest
@pytest.mark.parametrize(
    "role", [None, AdministratorRole.USERS, AdministratorRole.FULL]
)
async def test_create_administrator(spawn_client, fake2, snapshot, role, mongo):
    client = await spawn_client(authorize=True, administrator=True)

    user = await fake2.users.create()

    resp = await client.post("/administrators", {"user_id": user.id, "role": role})

    assert resp.status == 201

    if role == AdministratorRole.FULL:
        assert await get_one_field(mongo.users, "administrator", user.id) is True

    assert await resp.json() == snapshot


@pytest.mark.apitest
@pytest.mark.parametrize(
    "role", [AdministratorRole.USERS, AdministratorRole.FULL, AdministratorRole.BASE]
)
async def test_update_administrator(spawn_client, fake2, snapshot, role):
    client = await spawn_client(authorize=True, administrator=True)

    user = await fake2.users.create()

    authorization_client = client.app["authorization"]

    await authorization_client.add(
        AdministratorRoleAssignment(user.id, AdministratorRole.FULL),
    )

    resp = await client.patch(f"/administrators/{user.id}", {"role": role})

    assert resp.status == 200

    assert await resp.json() == snapshot


@pytest.mark.apitest
async def test_delete_administrator(spawn_client, fake2, snapshot, mongo):
    client = await spawn_client(authorize=True, administrator=True)

    user = await fake2.users.create()

    authorization_client = client.app["authorization"]

    await authorization_client.add(
        AdministratorRoleAssignment(user.id, AdministratorRole.BASE),
    )

    resp = await client.delete(f"/administrators/{user.id}")

    assert resp.status == 204

    assert await get_one_field(mongo.users, "administrator", user.id) is False

    assert await authorization_client.list_administrators() == [("test", AdministratorRole.FULL)]
