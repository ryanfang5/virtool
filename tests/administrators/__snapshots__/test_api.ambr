# name: TestUpdateUser.test
  <class 'dict'> {
    'active': True,
    'administrator': False,
    'administrator_role': None,
    'b2c': None,
    'b2c_display_name': None,
    'b2c_family_name': None,
    'b2c_given_name': None,
    'b2c_oid': None,
    'force_reset': True,
    'groups': <class 'list'> [
      <class 'dict'> {
        'id': 'bf1b993c',
        'name': 'musicians',
      },
    ],
    'handle': 'qtaylor',
    'id': '7cf872dc',
    'last_password_change': '2015-10-06T20:00:00Z',
    'permissions': <class 'dict'> {
      'cancel_job': False,
      'create_ref': False,
      'create_sample': False,
      'modify_hmm': False,
      'modify_subtraction': False,
      'remove_file': False,
      'remove_job': False,
      'upload_file': True,
    },
    'primary_group': <class 'dict'> {
      'id': 'bf1b993c',
      'name': 'musicians',
    },
  }
---
# name: TestUpdateUser.test_set_admin_roles[full-base-200]
  <class 'dict'> {
    'active': True,
    'administrator': False,
    'administrator_role': 'base',
    'b2c': None,
    'b2c_display_name': None,
    'b2c_family_name': None,
    'b2c_given_name': None,
    'b2c_oid': None,
    'force_reset': True,
    'groups': <class 'list'> [
      <class 'dict'> {
        'id': 'bf1b993c',
        'name': 'musicians',
      },
    ],
    'handle': 'qtaylor',
    'id': '7cf872dc',
    'last_password_change': '2015-10-06T20:00:00Z',
    'permissions': <class 'dict'> {
      'cancel_job': False,
      'create_ref': False,
      'create_sample': False,
      'modify_hmm': False,
      'modify_subtraction': False,
      'remove_file': False,
      'remove_job': False,
      'upload_file': True,
    },
    'primary_group': None,
  }
---
# name: TestUpdateUser.test_set_admin_roles[users-None-200]
  <class 'dict'> {
    'active': True,
    'administrator': False,
    'administrator_role': None,
    'b2c': None,
    'b2c_display_name': None,
    'b2c_family_name': None,
    'b2c_given_name': None,
    'b2c_oid': None,
    'force_reset': True,
    'groups': <class 'list'> [
      <class 'dict'> {
        'id': 'bf1b993c',
        'name': 'musicians',
      },
    ],
    'handle': 'qtaylor',
    'id': '7cf872dc',
    'last_password_change': '2015-10-06T20:00:00Z',
    'permissions': <class 'dict'> {
      'cancel_job': False,
      'create_ref': False,
      'create_sample': False,
      'modify_hmm': False,
      'modify_subtraction': False,
      'remove_file': False,
      'remove_job': False,
      'upload_file': True,
    },
    'primary_group': None,
  }
---
# name: test_get_roles
  <class 'list'> [
    <class 'dict'> {
      'description': 'Manage who is an administrator and what they can do.',
      'id': 'full',
      'name': 'Full',
    },
    <class 'dict'> {
      'description': 'Manage instance settings and administrative messages.',
      'id': 'settings',
      'name': 'Settings',
    },
    <class 'dict'> {
      'description': 'Manage users in any space. Delete any space.',
      'id': 'spaces',
      'name': 'Spaces',
    },
    <class 'dict'> {
      'description': 'Create user accounts. Control activation of user accounts.',
      'id': 'users',
      'name': 'Users',
    },
    <class 'dict'> {
      'description': '
        Provides ability to:
        - Create new spaces even if the `Free Spaces` setting is not enabled.
        - Manage HMMs and common references.
        - View all running jobs.
        - Cancel any job.
      ',
      'id': 'base',
      'name': 'Base',
    },
  ]
---
# name: test_get_user
  <class 'dict'> {
    'active': True,
    'administrator': False,
    'administrator_role': 'base',
    'b2c': None,
    'b2c_display_name': None,
    'b2c_family_name': None,
    'b2c_given_name': None,
    'b2c_oid': None,
    'force_reset': False,
    'groups': <class 'list'> [
    ],
    'handle': 'leeashley',
    'id': 'bf1b993c',
    'last_password_change': '2015-10-06T20:00:00Z',
    'permissions': <class 'dict'> {
      'cancel_job': False,
      'create_ref': False,
      'create_sample': False,
      'modify_hmm': False,
      'modify_subtraction': False,
      'remove_file': False,
      'remove_job': False,
      'upload_file': False,
    },
    'primary_group': None,
  }
---
# name: test_list_users
  <class 'dict'> {
    'found_count': 3,
    'items': <class 'list'> [
      <class 'dict'> {
        'active': True,
        'administrator': True,
        'administrator_role': 'full',
        'b2c': None,
        'b2c_display_name': None,
        'b2c_family_name': None,
        'b2c_given_name': None,
        'b2c_oid': None,
        'force_reset': False,
        'groups': <class 'list'> [
        ],
        'handle': 'bob',
        'id': 'test',
        'last_password_change': '2015-10-06T20:00:00Z',
        'permissions': <class 'dict'> {
          'cancel_job': False,
          'create_ref': False,
          'create_sample': False,
          'modify_hmm': False,
          'modify_subtraction': False,
          'remove_file': False,
          'remove_job': False,
          'upload_file': False,
        },
        'primary_group': None,
      },
      <class 'dict'> {
        'active': True,
        'administrator': False,
        'administrator_role': 'base',
        'b2c': None,
        'b2c_display_name': None,
        'b2c_family_name': None,
        'b2c_given_name': None,
        'b2c_oid': None,
        'force_reset': False,
        'groups': <class 'list'> [
        ],
        'handle': 'leeashley',
        'id': 'bf1b993c',
        'last_password_change': '2015-10-06T20:00:00Z',
        'permissions': <class 'dict'> {
          'cancel_job': False,
          'create_ref': False,
          'create_sample': False,
          'modify_hmm': False,
          'modify_subtraction': False,
          'remove_file': False,
          'remove_job': False,
          'upload_file': False,
        },
        'primary_group': None,
      },
      <class 'dict'> {
        'active': True,
        'administrator': False,
        'administrator_role': 'full',
        'b2c': None,
        'b2c_display_name': None,
        'b2c_family_name': None,
        'b2c_given_name': None,
        'b2c_oid': None,
        'force_reset': False,
        'groups': <class 'list'> [
        ],
        'handle': 'zclark',
        'id': 'fb085f7f',
        'last_password_change': '2015-10-06T20:00:00Z',
        'permissions': <class 'dict'> {
          'cancel_job': False,
          'create_ref': False,
          'create_sample': False,
          'modify_hmm': False,
          'modify_subtraction': False,
          'remove_file': False,
          'remove_job': False,
          'upload_file': False,
        },
        'primary_group': None,
      },
    ],
    'page': 1,
    'page_count': 1,
    'per_page': 25,
    'total_count': 3,
  }
---
# name: test_update_admin_role[None]
  <class 'dict'> {
    'active': True,
    'administrator': False,
    'administrator_role': None,
    'b2c': None,
    'b2c_display_name': None,
    'b2c_family_name': None,
    'b2c_given_name': None,
    'b2c_oid': None,
    'force_reset': False,
    'groups': <class 'list'> [
    ],
    'handle': 'leeashley',
    'id': 'bf1b993c',
    'last_password_change': '2015-10-06T20:00:00Z',
    'permissions': <class 'dict'> {
      'cancel_job': False,
      'create_ref': False,
      'create_sample': False,
      'modify_hmm': False,
      'modify_subtraction': False,
      'remove_file': False,
      'remove_job': False,
      'upload_file': False,
    },
    'primary_group': None,
  }
---
# name: test_update_admin_role[full]
  <class 'dict'> {
    'active': True,
    'administrator': True,
    'administrator_role': 'full',
    'b2c': None,
    'b2c_display_name': None,
    'b2c_family_name': None,
    'b2c_given_name': None,
    'b2c_oid': None,
    'force_reset': False,
    'groups': <class 'list'> [
    ],
    'handle': 'leeashley',
    'id': 'bf1b993c',
    'last_password_change': '2015-10-06T20:00:00Z',
    'permissions': <class 'dict'> {
      'cancel_job': False,
      'create_ref': False,
      'create_sample': False,
      'modify_hmm': False,
      'modify_subtraction': False,
      'remove_file': False,
      'remove_job': False,
      'upload_file': False,
    },
    'primary_group': None,
  }
---
# name: test_update_admin_role[users]
  <class 'dict'> {
    'active': True,
    'administrator': False,
    'administrator_role': 'users',
    'b2c': None,
    'b2c_display_name': None,
    'b2c_family_name': None,
    'b2c_given_name': None,
    'b2c_oid': None,
    'force_reset': False,
    'groups': <class 'list'> [
    ],
    'handle': 'leeashley',
    'id': 'bf1b993c',
    'last_password_change': '2015-10-06T20:00:00Z',
    'permissions': <class 'dict'> {
      'cancel_job': False,
      'create_ref': False,
      'create_sample': False,
      'modify_hmm': False,
      'modify_subtraction': False,
      'remove_file': False,
      'remove_job': False,
      'upload_file': False,
    },
    'primary_group': None,
  }
---
